### ISSUE TITLE
Code Intelligence 3.19 Tracking issue

### ISSUE THREAD

**Issue by efritz** [sourcegraph/sourcegraph#12132]
### Plan

**Goal 1: [Add support for monorepos](https://github.com/orgs/sourcegraph/projects/76)**

Unblock customers [C1](https://app.hubspot.com/contacts/2762526/company/464956351/) & [C2](https://app.hubspot.com/contacts/2762526/company/554275594/) from adopting LSIF by removing known blockers for large repositories with high commit rates. This milestone will make progress of removing the known blockers for these companies by following tasks (will still need a follow-up in (at least) the 3.20 milestone):

Building a prototype to speed up indexing on all repos. This is necessary for large monorepos as indexing the entire codebase is infeasible. This prototype will act as the feasibility study referenced in the [RFC 170](https://docs.google.com/document/d/1NPu0Vc7FpdoYwCrtpnu-8KB4OPbw7L0KBTqw96JVc8w). If we determine that this technique is viable, then the 3.20 milestone will focus on a proper implementation of the prototype (which also requires updating indexers to create patch indexes). @efritz @gbrik 

**Goal 2: Improve developer experience for use of LSIF in CI**

Users aren't configuring LSIF in CI because it is technically cumbersome, carries a maintenance burden, difficult to explain, and may involve red tape. We are aiming to alleviate these issues by automating the process and reduce number setup steps needed by developers to deploy. To support this effort we are also going to create additional quickstart documentation.

**Goal 3: Increase LSIF adoption**

Users are blocked from adopting LSIF because existing indexers don't support their use case, or there is no indexer for their language. We plan to distribute lsif-clang to customers that already have precise C++ support to ensure that it continues to meet their use case and is ready for wider distribution. We will also write a plan for precise SWIFT & Java support. 

**Goal 4: Resolve technical debt to improve quality and reliability**

This work is to improve overall quality of code intel components, including fixing worker concurrency bug, UI hover experience and implementing graceful failover for bundle manager.

### Availability

Period is from **2020-07-20** to **2020-08-20**. Please write the days you won't be working and the number of working days for the period.

- Aida: 24d
- Eric: 17d (out 07/23-07/24)
- Garo: 11d (out first week)

### Workload

@aidaeology

- [x] Determine state of LSIF-java [#12306](https://github.com/sourcegraph/sourcegraph/issues/12306) 
- [ ] ~Make bundle manager janitor more lenient during disaster recovery~ [#12168](https://github.com/sourcegraph/sourcegraph/issues/12168) 🧶
- [x] Research language usage among customers and projects [#12726](https://github.com/sourcegraph/sourcegraph/issues/12726) 
- [ ] ~Programatically create regular data about code intel activity~ [#11746](https://github.com/sourcegraph/sourcegraph/issues/11746) 
- [x] Update left side navigation [#13188](https://github.com/sourcegraph/sourcegraph/pull/13188) :shipit:
- [x] Edit of language guide page to improve phrasing [#13108](https://github.com/sourcegraph/sourcegraph/pull/13108) :shipit:
- [x] Updated adding lsif to workflows docs [#13078](https://github.com/sourcegraph/sourcegraph/pull/13078) :shipit:
- [x] Improving code intelligence docs and LSIF indexer installation instructions [#13041](https://github.com/sourcegraph/sourcegraph/pull/13041) :shipit:
- [x] Added goals to code intel team page [#1369](https://github.com/sourcegraph/about/pull/1369) :shipit:
- [x] Updated code intel page with NEO [#1327](https://github.com/sourcegraph/about/pull/1327) :shipit:

@efritz

- [x] Implement better nearest commit queries [#12098](https://github.com/sourcegraph/sourcegraph/issues/12098) 
- [x] Visible at tip calculation is racy [#12095](https://github.com/sourcegraph/sourcegraph/issues/12095) 🐛
- [x] Query definition and hover provider so we can correctly badge results [#12133](https://github.com/sourcegraph/sourcegraph/issues/12133) 
- [x] RFC 199: Index queue API [#12664](https://github.com/sourcegraph/sourcegraph/issues/12664) 
- [x] RFC 199: Index queue client [#12665](https://github.com/sourcegraph/sourcegraph/issues/12665) 
- [x] RFC 199: Expose internal routes for the indexer [#12666](https://github.com/sourcegraph/sourcegraph/issues/12666) 
- [x] RFC 199: Build indexer service [#12707](https://github.com/sourcegraph/sourcegraph/issues/12707) 
- [ ] ~RFC 199: Use firecracker~ [#12708](https://github.com/sourcegraph/sourcegraph/issues/12708) 
- [ ] RFC 199: Deploy new indexer service [#12709](https://github.com/sourcegraph/sourcegraph/issues/12709) 
- [ ] codeintel: Alert on failure to open database [#12712](https://github.com/sourcegraph/sourcegraph/issues/12712) 🧶
- [x] Create quick-start video for setting up Code Intelligence in CI/CD [#13049](https://github.com/sourcegraph/sourcegraph/issues/13049) 
- [ ] ~UI Tooltips for asymmetric precision of hover and definition results~ [#12706](https://github.com/sourcegraph/sourcegraph/issues/12706) 
- [ ] Remove -endpoint from src-cli usage [#11878](https://github.com/sourcegraph/sourcegraph/issues/11878) 🧶
- [x] codeintel: Install and use the LSIF upload route on the codeintel internal API [#13157](https://github.com/sourcegraph/sourcegraph/pull/13157) :shipit:
- [x] Add -upload-route flag to lsif upload [#267](https://github.com/sourcegraph/src-cli/pull/267) :shipit:
- [ ] Add hover alert for partial precise results [#459](https://github.com/sourcegraph/code-intel-extensions/pull/459) :shipit:
- [x] codeintel: Install docker inside precise-code-intel-indexer-vm docker image [#13119](https://github.com/sourcegraph/sourcegraph/pull/13119) :shipit:
- [x] docker-images: Add ignite-ubuntu [#12919](https://github.com/sourcegraph/sourcegraph/pull/12919) :shipit:
- [x] codeintel: Add fast-path edge unmarshalling [#12878](https://github.com/sourcegraph/sourcegraph/pull/12878) :shipit:🧶
- [x] codeintel: Add disable indexer flag [#12800](https://github.com/sourcegraph/sourcegraph/pull/12800) :shipit:
- [x] codeintel: Add mocks for queue client [#12798](https://github.com/sourcegraph/sourcegraph/pull/12798) :shipit:
- [x] internal API proxy: Remove verb allowlist in gitservice proxy [#12797](https://github.com/sourcegraph/sourcegraph/pull/12797) :shipit:
- [x] codeintel: Collapse worker handler and processor [#12795](https://github.com/sourcegraph/sourcegraph/pull/12795) :shipit:
- [x] workerutil: Make generic store [#12792](https://github.com/sourcegraph/sourcegraph/pull/12792) :shipit:
- [x] Fix default host for SRC_HTTP_ADDR_INTERNAL [#12768](https://github.com/sourcegraph/sourcegraph/pull/12768) :shipit:
- [x] Revert "Revert "codeintel: Internal API proxy (#12691)" (#12728)" [#12758](https://github.com/sourcegraph/sourcegraph/pull/12758) :shipit:
- [x] codeintel: VM-based indexer service [#12723](https://github.com/sourcegraph/sourcegraph/pull/12723) :shipit:
- [x] codeintel: Internal API proxy [#12691](https://github.com/sourcegraph/sourcegraph/pull/12691) :shipit:
- [x] codeintel: Index queue client [#12688](https://github.com/sourcegraph/sourcegraph/pull/12688) :shipit:
- [x] workerutil: Move store into own package [#12663](https://github.com/sourcegraph/sourcegraph/pull/12663) :shipit:
- [x] workerutil: Add DequeueWithIndependentTransactionContext to store [#12661](https://github.com/sourcegraph/sourcegraph/pull/12661) :shipit:
- [x] codeintel: Index queue API [#12657](https://github.com/sourcegraph/sourcegraph/pull/12657) :shipit:
- [x] Remove committed binary [#12594](https://github.com/sourcegraph/sourcegraph/pull/12594) :shipit:
- [x] Add index on repo name column [#12591](https://github.com/sourcegraph/sourcegraph/pull/12591) :shipit:
- [x] Categorize enterprise frontend startup behaviors [#12539](https://github.com/sourcegraph/sourcegraph/pull/12539) :shipit:
- [x] Move [enterprise/]cmd/frontend/authz to [enterprise/]internal/authz [#12538](https://github.com/sourcegraph/sourcegraph/pull/12538) :shipit:
- [x] codeintel: Remove ErrMalformedBundle [#12497](https://github.com/sourcegraph/sourcegraph/pull/12497) :shipit:
- [x] codeintel: Remove unused store, gitserver code [#12425](https://github.com/sourcegraph/sourcegraph/pull/12425) :shipit:
- [x]  codeintel: Update store to target new nearest upload tables [#12422](https://github.com/sourcegraph/sourcegraph/pull/12422) :shipit:
- [x] codeintel: Add commit updater utility [#12411](https://github.com/sourcegraph/sourcegraph/pull/12411) :shipit:
- [x] codeintel: Add CalculateVisibleUploads to store [#12408](https://github.com/sourcegraph/sourcegraph/pull/12408) :shipit:
- [x] workerutil: Fix nil-deref in when using worker store transactionally [#12407](https://github.com/sourcegraph/sourcegraph/pull/12407) :shipit:
- [x] codeintel: Add tables for denormalizing nearest upload data [#12406](https://github.com/sourcegraph/sourcegraph/pull/12406) :shipit:
- [x] codeintel: Add CommitGraph to gitserver client [#12404](https://github.com/sourcegraph/sourcegraph/pull/12404) :shipit:
- [x] codeintel: Add commit graph utilities [#12402](https://github.com/sourcegraph/sourcegraph/pull/12402) :shipit:
- [x] codeintel: Add Lock to store [#12401](https://github.com/sourcegraph/sourcegraph/pull/12401) :shipit:
- [x] workerutil: Make all column names customizable [#12398](https://github.com/sourcegraph/sourcegraph/pull/12398) :shipit:
- [x] codeintel: Apply rate limit to gitserver requests from indexability scheduler [#12379](https://github.com/sourcegraph/sourcegraph/pull/12379) :shipit:
- [x] codeintel: Group code intel data for serialization on-demand [#12125](https://github.com/sourcegraph/sourcegraph/pull/12125) :shipit:
- [x] codeintel: Additional worker memory improvements [#12108](https://github.com/sourcegraph/sourcegraph/pull/12108) :shipit:

@gbrik

- [ ] ~🚚 LSIF-clang Delivery~ [#12349](https://github.com/sourcegraph/sourcegraph/issues/12349) 
- [ ] ~Create a code intel user survey~ [#11747](https://github.com/sourcegraph/sourcegraph/issues/11747) 
- [ ] ~update lsif.dev~ [#12501](https://github.com/sourcegraph/sourcegraph/issues/12501) 
- [ ] ~Investigate effort for a new LSIF-swift indexer~ [#12350](https://github.com/sourcegraph/sourcegraph/issues/12350) 🕵️

#### Legend

- 👩 Customer issue
- 🐛 Bug
- 🧶 Technical debt
- 🛠️ [Roadmap](https://docs.google.com/document/d/1cBsE9801DcBF9chZyMnxRdolqM_1c2pPyGQz15QAvYI/edit#heading=h.5nwl5fv52ess)
- 🕵️ [Spike](https://en.wikipedia.org/wiki/Spike_(software_development))
- 🔒 Security issue
- :shipit: Pull Request

**Comment by efritz** [comment]
*2020/07/20-22*

Planning to implement https://github.com/sourcegraph/sourcegraph/issues/12098 to provide better nearest commit queries. This is basically calculating visibility of uploads from each commit in a repository ahead of time so that it is very fast on the query/hot path side.

After beginning to implement it, I see that it will diverge a bit from the RFC (just in some nuanced details and a few small oversights; the spirit is still the same). I don't foresee any major problems.

**Comment by efritz** [comment]
**Completed 2020/07/20-22**

Implemented an improved system for determining the nearest upload to a commit, which is nearly complete (waiting on reviews before merging).

**Completed 2020/07/27-31**

This week I plan to get [RFC 199: Arbitrary code execution in the auto-indexer](https://docs.google.com/document/d/1rCduWqaLAbMu2s43RwJTBbRlhL6qS3oqq4iawiGdoVE) approved and begin sketching an implementation. I will also dedicate time with @gbrik to start pairing on building out a proof of concept implementation for incremental indexing. The actual break-up of work here still needs discussion.

**Comment by efritz** [comment]
**Completed 2020/07/27-31**

Began work in splitting the indexer service into an API which controls the index queue and a worker that will run on a bare compute node and spawn firecracker VMs as described in [RFC 199](https://docs.google.com/document/d/1rCduWqaLAbMu2s43RwJTBbRlhL6qS3oqq4iawiGdoVE). This work is tracked in the draft PR https://github.com/sourcegraph/sourcegraph/pull/12537.

This will allow us to easily run arbitrary code required for TypeScript/Javascript dependency resolution, as well as enabling future work in [RFC 201](https://docs.google.com/document/d/1NPQs1s814LZjNXjPuavqC1N7hZR192DNtmSBmAeH9UY) to allow indexing projects with arbitrary configuration.

**Planned 2020/08/01-05**

Test and break apart the code in the draft PR created the previous week for review. The goal is to have it reviewed and merged by end of week.

Time allowing, I will begin work on exploratory work to determine how to gather metrics and logs for this service (will tag in @sourcegraph/distribution for ideas), and begin setting up the compute infrastructure to run the new indexer service. This task may be started, but will also likely be the focus of the following week.

**Comment by aidaeology** [comment]
**Completed 2020/07/31**
Finally finished all of the general & engineering onboarding tasks - local environment is up and running now.

Finished setting up and getting lsif-java to work locally - was able to reproduce all of the issues listed [here](https://github.com/sourcegraph/lsif-java/issues).  I was not able to get lsif-java to work for any projects built with Gradle following the configuration steps provided in the README.  I was able to generate LSIF dump for 4/10 OSS repositories (Maven projects only) and successfully uploaded to sourcegraph.com:
- [google/turbine](https://github.com/google/turbine)
- [square/javapoet](https://github.com/square/javapoet)
- [twitter/whiskey](https://github.com/twitter/whiskey)
- [json-iterator/java](https://github.com/json-iterator/java)

Conclusion: LSIF-java is nowhere near ready to ship.

**Planned 2020/08/03-05**
Worked with code intel team to create and complete our first code intel engineering onboarding plan, will follow this guide with Noah in two weeks.

Update the code-intel documentation for users to make it more user friendly and make it flow better.
Also continue to work on collecting user research for code intel.
Expect to start on Bundle Manager work near end of the week.

**Comment by aidaeology** [comment]
@gbrik gentle reminder to add your weekly update.

**Comment by gbrik** [comment]
### Completed 2020/07/31

Spent my heads down time on lsif-clang. Debugged Beyang's issue running the indexer on Ubuntu 18.04, but was unable to repro. Debugged user issue running the indexer in their dev environment. Narrowed down the cause, but can't attempt to directly repro so going to have to guess at the issue and try out fixes that way. Also can't directly repro the macos issue eric and mimi are hitting on my partner's mac, but it's a really old macbook air, so waiting on Aida to set up an old mac I can ssh into to test.

Other time was spent in meetings/code review. More than half my time was spent on this, which is not necessarily bad, just want to track it since I think balancing this has been the hardest part of working part time.

### Planned 2020/08/03

Make as much progress as possible on lsif-clang: poke beyang, research user issue, debug in the macos ssh env when it's set up.

Make code intel survey releasable (even if we don't actually push it through any channels yet): create a draft google survey, get a round of feedback on it, make a final survey. 

Start the incremental indexing prototype with eric: schedule a time to pair on this.

**Comment by efritz** [comment]
**Completed 2020/08/03-07**

Finished implementation of an API that exposes the auto-index work queue, and a new service (deployable to a compute node outside of k8s) that will run docker images (soon Firecracker VMs) to allow extensible and setup steps and indexer logic during auto-indexing. See the [task tracking issue](https://github.com/sourcegraph/sourcegraph/issues/12307) for details.

Had discussions with @daxmc99, @pecigonzalo, @slimsag, and @chayim to talk about deploying the new indexer service. Discussion notes are present in the [RFC 199: User code execution in the auto-indexer](https://docs.google.com/document/d/1rCduWqaLAbMu2s43RwJTBbRlhL6qS3oqq4iawiGdoVE).

TL;DR:

- We are going to create a new GCP project so that we have greater isolation between services (we will not need to firewall off subnets, or worry about users gaining access to the GCP API from within an index container). This will also allow us to see a cost breakdown more naturally for these compute resources.
- We are going to create a terraform config that will set up the node with Weavework Ignite, Firecracker, Docker, etc. This will run the indexer agent as a docker image so that renovate can update the image tag and re-run terraform so we get continuous deployment on this project.
- We are going to use stackdriver/built-in GCP logs for now.
- We are going to expose metrics from the indexer agents via public DNS (nothing here is private data). This will allow us to scrape the metrics from the Prometheus server on the cloud cluster.

**Planned 2020/08/03-07**

Next week I am planning on primarily working through the tasks in the deployment discussion above.

**Comment by gbrik** [comment]
#### Completed 2020/08/03-07

Mainly made progress on incremental indexing this week, halfway to a prototype. Got pretty derailed by insomnia, travel, etc, hoping to make up some progress this weekend. No new fixes found for lsif-clang :(

#### Planned 2020/08/10-17

Will update pending progress this weekend, but mainly want to get a run of the incremental indexer, and make some lsif-clang fixes.

**Comment by marekweb** [comment]
Dear all,

This is your release captain speaking. 🚂🚂🚂

Branch cut for the **3.19 release is scheduled for tomorrow**.

Is this issue / PR going to make it in time? Please change the milestone accordingly.
When in doubt, reach out!

Thank you

**Comment by gbrik** [comment]
#### Completed 2020/08/10-14

Created rough draft of code intel survey, going to polish a bit before sharing for feedback Monday. 

Successfully ssh'd into Aida's macbook for lsif-clang debugging, poked Beyang about the Ubuntu bug, no new discoveries re customer issue.

Made a lot of progress on incremental indexing code. Mostly just need to make a couple changes to source modules I'm less familiar with and lsif-go for parallel compilation before I can test.

#### Planned 2020/08/17-21

Finish incremental PoC, do all the doc updates still assigned to me, send out code intel survey.

**Comment by efritz** [comment]
**Completed 2020/08/10-14**

Focused a lot of efforts on deploying the indexer service in https://github.com/sourcegraph/infrastructure/pull/2056. Going slowly because I'm new to Terraform and there are some outstanding permissions issues that @sourcegraph/distribution has been helping me along with. Wrote new code intel developer documentation in https://github.com/sourcegraph/sourcegraph/pull/12949.

**Planned 2020/08/17-20**

Continue effort on deploying the indexer service. Release lsif-go 1.0 and write a blog post in time for the release about the performance improvements.

**Comment by efritz** [comment]
**Completed 2020/08/17-20**

- Buttoned-up and released lsif-go v1.0.0. There was a report of increased memory usage which I started to triage.
- Waiting for reviews on blog post before publication: https://github.com/sourcegraph/about/pull/1410
- Was planning to try to finish up deploying the indexer service. Progress has been steady but a little slow. Have been finding small errors that take a deploy cycle to dotcom to fix (stuff like the proper route not being exposed to the correct API).
